### PR TITLE
Line Item List Price

### PR DESCRIPTION
**Pull request by ashkan18**
# Problem(s)
We introduced `offer` in #261 , with that change, calculating some total amounts got complicated. Generally the relation between an `offer` and `lineItem` is little strange. Offers are made on `order`, they end up defining the total amounts being paid to seller and buyer and should be used to calculate things like shipping and tax as well.

# Solution
It seems the easiest/cleanest solution was to limit the number of line items on offer orders to 1. This means offer orders can't have more than one line item.
The `total_amount_cents` of that line item is calculated by the last offer of the order.
We still want to know the original `list_price` for that order, for that to work, we now have `total_list_price_cents` on `orderType` which basically calculates that by sum of all `list_price * quantity` of `line_items` of the order.

# Deprecated Field
we renamed `price_cents` to `list_price_cents` on `LineItem` to be more accurate about what that field represents. 
From GraphQL perspective we kept `price_cents` around but deprecated it and for now it returns `list_price_cents` so no change needed in clients immediately.

cc: @artsy/analytics  

this addresses: https://artsyproduct.atlassian.net/browse/PURCHASE-578

### CODE REVIEW

File: app/events/order_event.rb [hunk 1]
```diff
@@ -47,7 +47,7 @@ def properties
 
   def line_item_detail(line_item)
     {
-      price_cents: line_item.price_cents,
+      price_cents: line_item.list_price_cents,
```
Comment (williardx) [inline app/events/order_event.rb]: Would it make sense to update this `price_cents` as well for consistency? That could be a separate PR since I know we'd have to update Pulse as well.

File: app/models/line_item.rb [hunk 2]
```diff
@@ -3,7 +3,20 @@ class LineItem < ApplicationRecord
...
   has_many :fulfillments, through: :line_item_fulfillments
 
+  before_create :validate_creation
+
   def total_amount_cents
-    price_cents * quantity
+    order.mode == Order::BUY ? total_list_price : order.last_offer&.amount_cents
+  end
+
+  def total_list_price
```
Comment (williardx) [inline app/models/line_item.rb]: Can we make this `total_list_price_cents` to be consistent with our other `foo_cents` fields?

File: app/services/order_total_updater_service.rb [hunk 3]
```diff
@@ -27,15 +26,18 @@ def update_totals!
...
 
-  def calculate_commission_cents
-    if @order.mode == Order::OFFER
-      @order.offer_total_cents * @commission_rate
-    else
-      set_commission_on_line_items
-      @order.line_items.map(&:commission_fee_cents).sum
+  def can_calculate?
+    case @order.mode
+    when Order::BUY then @order.line_items.all? { |li| li.list_price_cents.present? }
```
Comment (williardx) [inline app/services/order_total_updater_service.rb]: For a buy order would we ever have a case where `list_price_cents` was `nil`?

Comment (williardx) [reply app/services/order_total_updater_service.rb]: Ah, I see now that we were checking for that earlier.

File: spec/models/line_item_spec.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
+require 'rails_helper'
+
+describe LineItem, type: :model do
```
Comment (williardx) [inline spec/models/line_item_spec.rb]: 🥇

File: _schema.graphql [hunk 5]
```diff
@@ -191,7 +191,8 @@ type LineItem {
     last: Int
   ): FulfillmentConnection
   id: ID!
-  priceCents: Int!
+  listPriceCents: Int!
+  priceCents: Int! @deprecated(reason: "switch to use listPrice")
```
Comment (joeyAghion) [inline _schema.graphql]: Do you mean `listPriceCents`?

File: app/events/order_event.rb (hunk 1 above)
Comment (joeyAghion) [reply app/events/order_event.rb]: +1

We could start returning a separate `list_price_cents` field in the event now, then remove the deprecated one only once Pulse has switched to using it.

File: app/events/transaction_event.rb [hunk 6]
```diff
@@ -45,7 +45,7 @@ def properties
 
   def line_item_detail(line_item)
     {
-      price_cents: line_item.price_cents,
+      price_cents: line_item.list_price_cents,
```
Comment (joeyAghion) [inline app/events/transaction_event.rb]: Ditto.

File: app/graphql/types/line_item_type.rb [hunk 7]
```diff
@@ -3,12 +3,17 @@ class Types::LineItemType < Types::BaseObject
   graphql_name 'LineItem'
 
   field :id, ID, null: false
-  field :price_cents, Integer, null: false
+  field :price_cents, Integer, null: false, deprecation_reason: 'switch to use listPrice'
```
Comment (joeyAghion) [inline app/graphql/types/line_item_type.rb]: Again... `listPriceCents`? `list_price_cents`?

File: app/graphql/types/order_type.rb [hunk 8]
```diff
@@ -30,6 +30,7 @@ class Types::OrderType < Types::BaseObject
     argument :from_id, String, required: false
     argument :from_type, String, required: false
   end
+  field :total_list_price_cents, Integer, null: false
```
Comment (joeyAghion) [inline app/graphql/types/order_type.rb]: Can't this be `null` for orders on works without buy-now enabled?

File: app/models/line_item.rb [hunk 9]
```diff
@@ -3,7 +3,20 @@ class LineItem < ApplicationRecord
   has_many :line_item_fulfillments, dependent: :destroy
   has_many :fulfillments, through: :line_item_fulfillments
 
+  before_create :validate_creation
```
Comment (joeyAghion) [inline app/models/line_item.rb]: You can make this a true validation with `on: :create` rather than a before hook that raises an exception.

```
validate :offer_order_lacks_line_items, on: :create

...

def offer_order_lacks_line_items
  errors.add(:order, "...") if order.mode == Order::OFFER && order.line_items.any?
end
```

File: app/services/sales_tax_service.rb [hunk 10]
```diff
@@ -75,7 +75,7 @@ def fetch_sales_tax
     @tax_client.tax_for_order(
       construct_tax_params(
         line_items: [{
-          unit_price: UnitConverter.convert_cents_to_dollars(@line_item.price_cents),
+          unit_price: UnitConverter.convert_cents_to_dollars(@line_item.total_amount_cents / @line_item.quantity),
```
Comment (joeyAghion) [inline app/services/sales_tax_service.rb]: It's weird to be redoing (or _undoing_) this calculation here. Line item should expose this as a method (like `effective_price_cents`).

File: app/models/line_item.rb [hunk 11]
```diff
@@ -3,7 +3,20 @@ class LineItem < ApplicationRecord
...
+
+  def total_list_price
+    list_price_cents * quantity
+  end
+
+  private
+
+  def validate_creation
+    # Offer orders can have only one line item
+    raise Errors::ValidationError, :offer_more_than_one_line_item if order.mode == Order::OFFER && order.line_items.exists?
```
Comment (yuki24) [inline app/models/line_item.rb]: Here `LineItem` and `Order` are circulating. Is there any reason this validation can't live in `Order`? Also, the `#exists?` method may query against the DB that may return `0` on creation. I think the `with_options` DSL may be helpful:

```ruby
class Order < ApplicationRecord
  ...

  # Common validations
  validates ...

  with_options if: -> { self.mode == Order::BUY } do
    # validations for Buy Now come here
    ...
  end

  with_options if: -> { self.mode == Order::OFFER } do
    # validations for Make Offer come here
    ...
    validates :line_items, length: { is: 1 }, on: :create # No need for a custom validation method as there's a built-in length validator
  end
end
```

File: app/graphql/types/line_item_type.rb (hunk 7 above)
Comment (ashkan18) [reply app/graphql/types/line_item_type.rb]: hehe the original one above is the autogenerated schema file :) 👍

File: app/graphql/types/order_type.rb (hunk 8 above)
Comment (ashkan18) [reply app/graphql/types/order_type.rb]: i know currently for make offer works, they should have list price too, but we can remove this restriction too.

File: app/models/line_item.rb (hunk 11 above)
Comment (ashkan18) [reply app/models/line_item.rb]: interesting! didn't know about `with_options`!! seems really useful! 

In this case, `length` seems to not be useful, it does a `length` on String and doesn't seem to work for arrays.
Also another question, if we create a `line_item` by lets say `Fabricate(:line_item, order: <some order>)` does this actually get invoked?

**Comment by ashkan18** [comment]
going to merge on green this one so I can test the whole path on staging with MP and Reaction PRs coming after

File: _schema.graphql [hunk 12]
```diff
@@ -278,6 +279,8 @@ type Order {
   currencyCode: String!
   displayCommissionRate: String
   id: ID!
+
+  # Item total in cents, for Offer Orders this field reflects current offer
   itemsTotalCents: Int!
```
Comment (sweir27) [inline _schema.graphql]: Wondering what the reasoning is for having this field be non-nullable.

Currently, if you create an "offer"-mode `Order`, it will initially have a `null` `itemsTotalCents`, until you add an initial offer to it. This seems valid so I'm wondering if it would be 👍 to relax this requirement.
